crates/metamask/src/password.rs

```diff
@@ -0,0 +1,190 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+/// Rust port of Metamask's password encryption/decryption logic.
+/// https://github.com/MetaMask/browser-passworder/blob/a8574c40d1e42b2bc2c2b3d330b0ea50aa450017/src/index.ts
+
+/// Inspired by:
+/// https://github.com/fedimint/fedimint/blob/aa21c66582c17a68f19438366864652cba4bd590/crypto/aead/src/lib.rs#L25
+/// https://docs.rs/ring/latest/ring/pbkdf2/index.html
+use crate::types::Vault;
+use aes_gcm::{
+    aead::{Aead, KeyInit, OsRng},
+    aes::{cipher::consts::U16, Aes256},
+    AesGcm, Nonce,
+};
+use base64::{engine::general_purpose, Engine as _};
+use pbkdf2::{hmac::Hmac, pbkdf2};
+use rand::{thread_rng, Rng, RngCore};
+use sha2::Sha256;
+use std::{error::Error, str};
+
+/// Nonce size is set at 16 bytes (128 bits).
+/// From:
+/// https://github.com/MetaMask/browser-passworder/blob/a8574c40d1e42b2bc2c2b3d330b0ea50aa450017/src/index.ts#L83
+pub type Aes256Gcm = AesGcm<Aes256, U16>;
+
+/// Encrypts a message using a key.
+///
+/// From:
+/// https://github.com/MetaMask/browser-passworder/blob/a8574c40d1e42b2bc2c2b3d330b0ea50aa450017/src/index.ts#L32
+pub fn encrypt(
+    password: &str,
+    data: &mut Vec<u8>,
+    key: Option<&[u8]>,
+    salt: Option<&str>,
+) -> Result<String, Box<dyn Error>> {
+    // Generate a salt if one is not provided.
+    let salt = salt.map_or_else(|| None, |s| Some(general_purpose::STANDARD.encode(s)));
+    let k = key_from_password(password, salt.as_ref().map(|s| s.as_bytes()));
+    let key = key.unwrap_or(&k);
+
+    // Generate the nonce (iv) from random bytes.
+    let mut rng = OsRng;
+    let mut bytes = [0u8; 16];
+    rng.fill_bytes(&mut bytes);
+    let nonce = Nonce::from_slice(&bytes);
+
+    // Construct a key from the provided bytes.
+    let cipher = Aes256Gcm::new(key.into());
+
+    // Encrypt the data.
+    let data = cipher.encrypt(nonce, data.as_ref()).unwrap();
```

Comment:
Not a fan of the unwrap unless it's unreachable -- maybe we can do something along the lines of `let data = cipher.encrypt(nonce, data.as_ref()).map_err(|e| e.to_string())?;`?

Reply:
Ahhhhhhhh this makes so much sense. Thank you so much. Jesus.